File: commands/uninstall.ts
import { Command } from 'commander';
import fs from 'fs';
import os from 'os';
import path from 'path';
import { execFileSync } from 'child_process';
import { SimpleGit } from 'simple-git';
import { AbstractCommand } from './base';

export class UninstallCommand extends AbstractCommand {
  constructor(protected git: SimpleGit) { super(git); }

  public load(program: Command): void {
    program
      .command('uninstall')
      .description('Remove wt shell completions and print package uninstall instructions')
      .option('--shell <shell>', 'Shell name (zsh|fish|all)')
      .option('--yes', 'Do not prompt (currently no prompts)')
      .action(async (options: { shell?: string }) => {
        await this.uninstall(options);
      });
  }

  private detectShell(): string {
    if (process.env.ZSH_VERSION) return 'zsh';
    if (process.env.FISH_VERSION) return 'fish';
    if (process.env.BASH_VERSION) return 'bash';

    const env = process.env.SHELL;
    if (env) return path.basename(env);
    return 'zsh';
  }

  private resolveWtBinaryPath(): string | null {
    try {
      const out = execFileSync('which', ['wt'], { encoding: 'utf8' });
      const p = out.trim();
      return p.length > 0 ? p : null;
    } catch {
      // ignore
    }

    try {
      const out = execFileSync('whereis', ['-b', 'wt'], { encoding: 'utf8' });
      const tokens = out
        .trim()
        .split(/\s+/)
        .map((s) => s.trim())
        .filter(Boolean);

      const paths = tokens.filter((t) => t.includes('/'));
      if (paths.length === 0) return null;
      return paths[0] ?? null;
    } catch {
      return null;
    }
  }

  private uninstallCommandHint(): string {
    const ua = process.env.npm_config_user_agent ?? '';
    if (ua.includes('pnpm')) return 'pnpm remove -g wt';
    if (ua.includes('yarn')) return 'yarn global remove wt';
    if (ua.includes('bun')) return 'bun remove -g wt';
    return 'npm uninstall -g wt';
  }

  private uninstallCommandHintsForBinary(binaryPath: string | null): string[] {
    const hints: string[] = [];

    if (binaryPath) {
      if (binaryPath.includes('/opt/homebrew/') || binaryPath.includes('/usr/local/')) {
        hints.push('brew uninstall wt');
      }
      if (binaryPath.includes('/Library/pnpm/') || binaryPath.includes('/pnpm/')) {
        hints.push('pnpm remove -g wt');
      }
    }

    hints.push(this.uninstallCommandHint());

    return Array.from(new Set(hints));
  }

  private uninstallZshCompletion(): void {
    const home = os.homedir();
    const completionFile = path.join(home, '.zsh', 'completions', '_wt');
    const zshrcPath = path.join(home, '.zshrc');

    if (fs.existsSync(completionFile)) {
      try {
        fs.unlinkSync(completionFile);
        this.log(`removed ${completionFile}`);
      } catch {
        this.log(`failed to remove ${completionFile}`);
      }
    }

    const startMarker = '# wt completion start';
    const endMarker = '# wt completion end';

    if (!fs.existsSync(zshrcPath)) return;

    try {
      const existing = fs.readFileSync(zshrcPath, 'utf8');
      const start = existing.indexOf(startMarker);
      const end = existing.indexOf(endMarker);

      if (start === -1 || end === -1 || end < start) return;

      const afterEnd = existing.indexOf('\n', end);
      const cutEnd = afterEnd === -1 ? existing.length : afterEnd + 1;

      const next = (existing.slice(0, start) + existing.slice(cutEnd)).replace(/\n{3,}/g, '\n\n');
      fs.writeFileSync(zshrcPath, next, { encoding: 'utf8' });
      this.log(`updated ${zshrcPath}`);
    } catch {
      this.log(`failed to update ${zshrcPath}`);
    }
  }

  private uninstallFishCompletion(): void {
    const home = os.homedir();
    const completionFile = path.join(home, '.config', 'fish', 'completions', 'wt.fish');

    if (!fs.existsSync(completionFile)) return;

    try {
      fs.unlinkSync(completionFile);
      this.log(`removed ${completionFile}`);
    } catch {
      this.log(`failed to remove ${completionFile}`);
    }
  }

  private async uninstall(options: { shell?: string }): Promise<void> {
    const shell = (options.shell ?? this.detectShell()).toLowerCase();

    if (shell === 'zsh') {
      this.uninstallZshCompletion();
    } else if (shell === 'fish') {
      this.uninstallFishCompletion();
    } else if (shell === 'all') {
      this.uninstallZshCompletion();
      this.uninstallFishCompletion();
    } else {
      console.error('Only zsh, fish, and all are supported');
      process.exit(1);
    }

    const binary = this.resolveWtBinaryPath();
    if (binary) {
      process.stdout.write(`wt binary:\n- ${binary}\n`);
    }

    process.stdout.write('Uninstall package (if installed globally):\n');
    for (const hint of this.uninstallCommandHintsForBinary(binary)) {
      process.stdout.write(`- ${hint}\n`);
    }
  }
}

File: src/cmd/uninstall.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

use anyhow::Result;
use clap::Args;

use crate::output;

#[derive(Args, Debug)]
pub struct UninstallArgs {
    /// Shell name (zsh|fish|bash|all)
    #[arg(long)]
    shell: Option<String>,

    /// Do not prompt
    #[arg(long)]
    yes: bool,
}

pub fn run(args: UninstallArgs) -> Result<()> {
    let shell = args
        .shell
        .unwrap_or_else(detect_shell)
        .to_lowercase();

    match shell.as_str() {
        "zsh" => uninstall_zsh_completion(),
        "fish" => uninstall_fish_completion(),
        "bash" => uninstall_bash_completion(),
        "all" => {
            uninstall_zsh_completion();
            uninstall_fish_completion();
            uninstall_bash_completion();
        }
        _ => {
            eprintln!("Only zsh, fish, bash, and all are supported");
            std::process::exit(1);
        }
    }

    if let Some(binary) = resolve_wt_binary() {
        println!("wt binary:\n- {}", binary);
    }

    println!("Uninstall package (if installed globally):");
    println!("- cargo uninstall wt-cli");
    Ok(())
}

fn detect_shell() -> String {
    if std::env::var("ZSH_VERSION").is_ok() { return "zsh".into(); }
    if std::env::var("FISH_VERSION").is_ok() { return "fish".into(); }
    if std::env::var("BASH_VERSION").is_ok() { return "bash".into(); }
    std::env::var("SHELL")
        .ok()
        .and_then(|s| {
            std::path::Path::new(&s)
                .file_name()
                .and_then(|n| n.to_str())
                .map(|s| s.to_string())
        })
        .unwrap_or_else(|| "zsh".into())
}

fn resolve_wt_binary() -> Option<String> {
    Command::new("which")
        .arg("wt")
        .output()
        .ok()
        .and_then(|o| {
            let s = String::from_utf8_lossy(&o.stdout).trim().to_string();
            if s.is_empty() { None } else { Some(s) }
        })
}

fn uninstall_zsh_completion() {
    let home = dirs::home_dir().unwrap();
    let completion_file = home.join(".zsh").join("completions").join("_wt");
    let zshrc = home.join(".zshrc");

    if completion_file.exists() {
        if fs::remove_file(&completion_file).is_ok() {
            output::log(&format!("removed {}", completion_file.display()));
        }
    }

    remove_block_from_file(&zshrc, "# wt completion start", "# wt completion end");
}

fn uninstall_fish_completion() {
    let home = dirs::home_dir().unwrap();
    let completion_file = home
        .join(".config")
        .join("fish")
        .join("completions")
        .join("wt.fish");

    if completion_file.exists() {
        if fs::remove_file(&completion_file).is_ok() {
            output::log(&format!("removed {}", completion_file.display()));
        }
    }
}

fn uninstall_bash_completion() {
    let home = dirs::home_dir().unwrap();
    let completion_file = home.join(".bash_completion.d").join("wt");
    let bashrc = home.join(".bashrc");

    if completion_file.exists() {
        if fs::remove_file(&completion_file).is_ok() {
            output::log(&format!("removed {}", completion_file.display()));
        }
    }

    remove_block_from_file(&bashrc, "# wt completion start", "# wt completion end");
}

fn remove_block_from_file(path: &PathBuf, start_marker: &str, end_marker: &str) {
    if !path.exists() { return; }
    let Ok(content) = fs::read_to_string(path) else { return; };
    let start = content.find(start_marker);
    let end = content.find(end_marker);
    if let (Some(s), Some(e)) = (start, end) {
        if e < s { return; }
        let cut_end = content[e..].find('\n').map_or(content.len(), |i| e + i + 1);
        let next = format!("{}{}", &content[..s], &content[cut_end..])
            .replace("\n\n\n", "\n\n");
        if fs::write(path, &next).is_ok() {
            output::log(&format!("updated {}", path.display()));
        }
    }
}
